radicle-surf/src/git/repo.rs

```diff
@@ -249,12 +257,72 @@ impl Repository {
             .map_err(|err| Error::ToCommit(err.into()))?;
         let git2_commit = self.inner.find_commit((commit.id).into())?;
         let tree = git2_commit.as_object().peel_to_tree()?;
-        Ok(Directory::root(tree.id().into()))
+        Ok(Directory::root(tree.id().into(), commit))
+    }
+
+    /// Returns a [`Directory`] for `path` in `commit`.
+    pub fn directory<C: ToCommit, P: AsRef<Path>>(
+        &self,
+        commit: C,
+        path: &P,
+    ) -> Result<Directory, Error> {
+        let root = self.root_dir(commit)?;
+        root.find_directory(path, self)?
+            .ok_or_else(|| Error::PathNotFound(path.as_ref().to_path_buf()))
+    }
+
+    /// Returns a [`File`] for `path` in `commit`.
+    pub fn file<C: ToCommit, P: AsRef<Path>>(&self, commit: C, path: &P) -> Result<File, Error> {
+        let root = self.root_dir(commit)?;
+        root.find_file(path, self)?
+            .ok_or_else(|| Error::PathNotFound(path.as_ref().to_path_buf()))
+    }
+
+    /// Returns a [`Tree`] for `path` in `commit`.
+    pub fn tree<C: ToCommit, P: AsRef<Path>>(&self, commit: C, path: &P) -> Result<Tree, Error> {
+        let commit = commit
+            .to_commit(self)
+            .map_err(|e| Error::ToCommit(e.into()))?;
+        let dir = self.directory(commit.id, path)?;
+        let mut entries = dir
+            .entries(self)?
+            .map(|en| {
+                let name = en.name().to_string();
+                let path = en.path();
+                let commit = self
+                    .last_commit(&path, commit.id)?
+                    .ok_or(Error::PathNotFound(path))?;
+                let commit_header = Header::from(commit);
+                Ok(TreeEntry::new(name, en.into(), commit_header))
+            })
+            .collect::<Result<Vec<TreeEntry>, Error>>()?;
```

Comment:
I don't think the `TreeEntry` or `Error` are necessary, not a blocker.
```suggestion
            .collect::<Result<Vec<_>, _>>()?;
```

Reply:
It didn't work for me.  I would get the following errors: 
```
error[E0282]: type annotations needed
   --> radicle-surf/src/git/repo.rs:296:17
    |
296 |                 Ok(TreeEntry::new(name, en.into(), commit_header))
    |                 ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
    |
help: consider specifying the generic arguments
    |
296 |                 Ok::<object::tree::TreeEntry, E>(TreeEntry::new(name, en.into(), commit_header))
    |                   ++++++++++++++++++++++++++++++

error[E0283]: type annotations needed
   --> radicle-surf/src/git/repo.rs:296:17
    |
296 |                 Ok(TreeEntry::new(name, en.into(), commit_header))
    |                 ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
297 |             })
298 |             .collect::<Result<Vec<_>, _>>()?;
    |                                            - type must be known at this point
```

Reply:
stoopid rust